style(vars): FRMW-546 Setup Global Vars

ssalinas24 opened the pull request:
JIRA: https://jira.rax.io/browse/FRMW-546

- [x] Dev LGTM
- [x] Dev LGTM
- [x] Design LGTM

I have added variables that I believe will help consolidate and unify some of the duplicated properties and styling that i have seen through out components. However, this is only based on my brief history with the framework, and the items that I have had related tasks to. I would highly appreciate any feedback from any DEVS that have way more history with the framework, in case I can add additional items that i may have missed.

ssalinas24 commented:
Some updates to vars for Kebab Case have also been implemented since they related to the sections that I was in.

ssalinas24 commented:
@glynnis If you agree with items that I have added, then you may want to have Anu look at the TODO comments in the code and create stories when it looks like this PR is going to complete.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
 
 // # Text colors
-@disabled-text-color: @gray-600;
+@headers-text-color: @gray-900; // TODO: Once Approved,then Implement Into Components
```
I would change this to the singular, `@header-text-color`

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@disabled-text-color: @gray-600;
+@disabled-icon-color:@gray-600;
+@link-color: @blue-700;
+@link-hover-color: saturate(@link-color, 15%);
```
We should double check with @ElementE that this is how he wants hover to appear. I know we've talked about other color formulas in other areas, but running this particular one by him can't hurt since links are everywhere :smile:

glynnis replied:
I spoke with @ElementE and `saturate` and the way it's defined here is indeed what we want :+1:

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+
+// # App Default Borders
+// TODO: Once Approved,then Define and Implement Into Components.
+@app-default-border-color: @gray-400;
```
Where does this get used? It's easy to tell that it's a border color from the var name, but not sure I know offhand where this gets used.

This one also makes me think about the `@app-` prefix -- do you think things like `@link-color` and `@disabled-text-color` should also get an `@app` prefix? Basically: what determines when variables get `@app-` and when they don't? Not sure I have the right answer, but something worth thinking about.

ssalinas24 replied:
@glynnis This is a new var that I was suggesting. I started remember about all the components, and inputs that I had run across with borders. Most of which were using the `@gray-400`. Thus, I was thinking we could benefit from an global app variable. This has not been implemented anywhere yet.

glynnis replied:
Ah, ok. Thanks for the refresher :smile: Carry on :+1:

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-default-row-margin:0;
+@app-default-row-padding:0;
+@app-default-column-margin:0;
+@app-default-column-padding:0;
```
Please add 1 space between the colons and values here.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
-@linkColor: @blue-700; // *TODO* convert var name to kebab?
-@linkColorHover: saturate(@linkColor, 15%); // *TODO* convert var name to kebab?
+@disabled-text-color: @gray-600;
+@disabled-icon-color:@gray-600;
```
Please add 1 space between the colon and value here.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-default-selected-background-color: @blue-700;
+@app-default-selected-text-color: @white;
+@app-default-hover-background-color: @blue-100;
+@app-default-hover-text-color: @gray-900;
```
:+1: I think this batch of variables is especially useful so that we're able to make changes in just one place in the future if we want to tweak selected/hover states.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@info-alert-icon-color: @gray-950;
+@disabled-alert-background-color: @disabled-background-color;
+@disabled-alert-text-color: @disabled-text-color;
+@diabled-alert-icon-color: @disabled-icon-color;
```
`@diabled` -> `@disabled`

glynnis replied:
Should these variable names start with `@alert` since that's the common element? For example:
`@alert-active-background-color` and `@alert-pending-background-color`.

Starting with `@alert` seems to be more in line with how we've been naming/organizing vars for components -- by starting with the common element (`@rxButton-default-background-color` etc.).

CITguy replied:
Steven and I discussed this as well and we found that it makes more sense when the variable reads like a grammatically proper sentence fragment. In this case, `active` (adj.) describes `alert` (n.), so it is placed in front.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+// After reviewing Bootstrap, I learned that we can probably improve on our font vars and structure
+// So I added the following Below. A Dev/User could update these in this one place.
+// TODO: Once Approved, then Implement Into Components. COULD BE LARGE EPIC FOR A SPRINT
+@app-font: "Roboto", sans-serif;
```
I think we could also use a variable for our mono font -- `@app-mono-font`. I know we've brought it into the Framework for our code blocks, but we could expose it here as a var as well.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-font-size: 13px;
+@app-font-family-sans-serif: "Helvetica Neue", Helvetica, Arial, sans-serif;
+@app-font-family-serif: Georgia, "Times New Roman", Times, serif;
+@app-font-family-monospace: Menlo, Monaco, Consolas, "Courier New", monospace;
```
I think our monospace font is actually Roboto Mono, but listing the others you have here as fallbacks is :+1:

CITguy replied:
Correct, "Roboto Mono" (https://github.com/rackerlabs/encore-ui/blob/master/src/components/rxApp/common.less#L5)

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-line-height: 1.2;
+@app-font-size: 13px;
+@app-font-family-sans-serif: "Helvetica Neue", Helvetica, Arial, sans-serif;
+@app-font-family-serif: Georgia, "Times New Roman", Times, serif;
```
We should run this by @ElementE. I don't think we use serifs anywhere currently, so it's worth discussing whether we want this defined at all. But if it is useful to define, we might need to consult with him on what the preferred serif font is (it'll have to come from Google or some other open source font supplier, not Typekit) vs. what order fallbacks should be listed in.

glynnis replied:
Confirmed with @ElementE that you should remove the serif font definition and var. 

He'd also like to eliminate the `@app-font-family-sans-serif` var, and instead just list those fonts as fallbacks for `@app-font` like so:
```
@app-font: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif;
```

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-font: "Roboto", sans-serif;
+@app-line-height: 1.2;
+@app-font-size: 13px;
+@app-font-family-sans-serif: "Helvetica Neue", Helvetica, Arial, sans-serif;
```
What's the advantage of defining this here, separately from `@app-font`? I think I would prefer to delete `@app-font-family-sans-serif` and just keep `@app-font`. I think it's likely that @ElementE won't want to include any serif fonts either, so keeping just the one font var, `@app-font` makes the most sense, but let's run it by him to be sure.

glynnis replied:
@ElementE agrees that we could use a `@app-mono-font` definition.

ssalinas24 replied:
@glynnis this is a section that I started discussing with Eric, but would like to continue discussing with the both of you. I left `@app-font: "Roboto", sans-serif;` in place since it is being used currently. After looking at bootstrap, I noticed that we may benefit from dropping `@app-font` and potential enhance by adding some flexible vars.

Like having a sans serif and serif var defined in one place. Which are incorrect at the moment in this PR and should include Roboto and fallsbacks instead. Then we could link them into two new vars `@app-font-family-headers` and `@app-font-family-body`. These could be used through out the framework to target headers and body text individually. Thus, being more flexible and scalable then what we currently have.

glynnis commented on src/styles/vars.less:
```diff
@@ -76,6 +130,7 @@
 @preprodActiveBackground: #d41f26;
 
 // ## Tables
+// TODO: update these to 2.0 colors
```
Looks like these are already 2.0 colors, but could use updating to kebab.

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-font-family-serif: Georgia, "Times New Roman", Times, serif;
+@app-font-family-monospace: Menlo, Monaco, Consolas, "Courier New", monospace;
+@app-font-family-headers: @app-font-family-sans-serif;
+@app-font-family-body: @app-font-family-serif;
```
See https://github.com/rackerlabs/encore-ui/pull/1471/files#r48358227

glynnis commented on src/styles/vars.less:
```diff
@@ -55,20 +55,74 @@
+@app-font-family-sans-serif: "Helvetica Neue", Helvetica, Arial, sans-serif;
+@app-font-family-serif: Georgia, "Times New Roman", Times, serif;
+@app-font-family-monospace: Menlo, Monaco, Consolas, "Courier New", monospace;
+@app-font-family-headers: @app-font-family-sans-serif;
```
I would change this to `@app-header-font` (to more closely resemble `@app-font` and `@app-mono-font`. I like that you added this, since it's possible that in the future we'd want to use a different font for headers than what we use for body text :+1:

glynnis commented on src/styles/vars.less:
```diff
@@ -147,7 +202,7 @@
 /*
  * Buttons
- */
+ */ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
The wording in this comment isn't super clear to me. Do you mean something like, "Vars below should not be utilized outside the context of rxButton code unless otherwise noted"?

ssalinas24 replied:
All i meant was that I took time to check throughout the project for those vars, and they don't exist anywhere outside each component related files. Thus, not having any stragglers floating around and those section of vars only exist within their related components, as we would like. If i found any issues, then i added comments next to the var line.

The comment was just my check list signoff to remind us that I have checked that section/component.

glynnis replied:
:+1:

glynnis commented on src/styles/vars.less:
```diff
@@ -163,38 +218,40 @@
 /*
  * Button Groups (rxButton)
- */
+ */ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -163,38 +218,40 @@
-@rxForm-error-accent-color: @red-500;
-@rxForm-selected-accent-color: @blue-500;
-@rxForm-placeholder-color: #a9a9a9; // *TODO*
+ // TODO Vars Are Not Isolated To Component
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -163,38 +218,40 @@
 /*
  * rxRadio
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -270,6 +331,8 @@
  *  960  <= size < 1200    Tablet-Landscape
  *  1200 <= size           PC
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -278,6 +341,7 @@
 /*
  * rxMetadata
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -292,6 +356,7 @@
 /*
  * rxToggleSwitch
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -307,6 +372,7 @@
 /*
  * progressbar
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -330,12 +397,14 @@
 /*
  * rxBreadcrumbs
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented on src/styles/vars.less:
```diff
@@ -330,12 +397,14 @@
 /*
  * Modals
  */
+ // All Vars Are Isolated to The Component Files, Unless Commented Below
```
Ditto https://github.com/rackerlabs/encore-ui/pull/1471/files#r48359094

glynnis commented:
@ssalinas24 This is awesome work! Really feels like the culmination of a lot of organization/thinking around vars usage, and it's all coming together :clap: :clap: 

To your point, I think there's a lot here that we should spin off into other stories, but I didn't want to do that on my own while you and Eric are out, so I set up some time in the new year to walk through this carefully and create stories/resolve some of the questions I posted above.

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -55,20 +55,75 @@
+// # Notifications
+// These could be new Vars to sum up and consolidate rxNotify, rxSortColumn, and rxBulkSelect Warnings
+// TODO: Once Approved, then Implement Into Components
+@app-active-alert-background-color: @green-100;
```
@glynnis Based on Ryan's last comment, I kept this as `@app-active-alert-background-color` instead of something like `@app-alert-active-background-color`.

@CITguy last commented on this PR with the following (however I only see it in my email now):
"Steven and I discussed this as well and we found that it makes more sense when the variable reads like a grammatically proper sentence fragment. In this case, active (adj.) describes alert (n.), so it is placed in front."

glynnis replied:
:+1:

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -55,20 +55,75 @@
-@disabled-background-color: @gray-300;
-@linkColor: @blue-700; // *TODO* convert var name to kebab?
-@linkColorHover: saturate(@linkColor, 15%); // *TODO* convert var name to kebab?
+@app-headers-text-color: @gray-900; // TODO: Once Approved,then Implement Into Components
```
@CITguy FYI, I mentioned to Glynnis that perhaps all global variables should be prefixed with "@app-" as shown in many of these lines. Sound good?

If so, then you will notice that some CSS comments are reminding my self to create stories for certain areas.

CITguy replied:
`@app-*` prefix :+1:

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -55,20 +55,75 @@
+@app-line-height: 1.2;
+@app-font-size: 13px;
+@app-sans-serif-font-family: "Roboto", Helvetica, Arial, sans-serif;
+@app-monospace-font-family: "Roboto Mono", Courier New, Courier, monospace;
```
@glynnis  and @ElementE  I have updated the font families and fallbacks as we have discussed.

glynnis replied:
:+1:

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -55,20 +55,75 @@
+// After reviewing Bootstrap, I learned that we can probably improve on our font vars and structure
+// So I added the following Below. A Dev/User could update these in this one place.
+// TODO: Once Approved, then Implement Into Components. COULD BE LARGE EPIC FOR A SPRINT
+@app-font: "Roboto", sans-serif;
```
This will eventually be deprecated once new stories are addressed by replacing `@app-font` with `@app-header-font-family` and `@app-body-font-family` where appropriate.

glynnis replied:
:+1:

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -163,38 +219,40 @@
+@rxForm-error-text-color: @red-700; // Used In: rxForm, and rxCharacterCount
+@rxForm-error-accent-color: @red-500; // Used In: Vars, and rxCheckbox
+@rxForm-selected-accent-color: @blue-500; // Used In: Vars
+@rxForm-placeholder-color: #a9a9a9; // *TODO* Convert Hex to 2.0 Var
```
@glynnis and @CITguy some of these may be able to have values that link to the new "app-" variables that I have introduced in this PR.

Another idea is that some of these vars that share across components should become global vars or link back to new global vars that we can create. @CITguy what would you think about that?

For example, `@rxForm-gutter` could be replaced or link to something new like `@app-form-gutter-width`.

CITguy replied:
I like the idea of finding common global variable values. Though, I think `@app-gutter-width` would be more appropriate, because it would apply to more than just forms.

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -356,6 +425,17 @@
 @red: #fc0f1d;
 
+// # Fonts
+// deprecated
```
This section is where i have kept some of the original global vars that have had recent name updates to follow the "app-" naming prefix.

ssalinas24 commented on src/styles/vars.less:
```diff
@@ -356,6 +425,17 @@
+// # Text Colors
+// deprecated
+@linkColor: @link-color;
+@linkColorHover: @link-hover-color;
```
These link color var values will be updated to equal new ones with "app-" prefix in separate story.

ssalinas24 replied:
Story has been created.

ssalinas24 commented:
Feedback has been implemented and i have updated this PR.

glynnis commented:
This LGTM. 

@ssalinas24 When you create JIRA stories (or work with Anu to do so) for some of the remaining work you identified (like updating references `@app-font`, etc.), can you post the links in a comment here?

Nice work with this!

CITguy commented on src/styles/vars.less:
```diff
@@ -240,13 +300,15 @@
-@rxFieldName-symbol-color: @red-500;
-@rxFieldName-font-size: 12px;
+ // Vars Are Not Isolated To Component
+@rxFieldName-color: @gray-800; // Used In: rxForm, and rxCharacterCount
```
This might be good as `@app-label-color`.

glynnis replied:
:+1:

CITguy commented on src/styles/vars.less:
```diff
@@ -240,13 +300,15 @@
-@rxFieldName-font-size: 12px;
+ // Vars Are Not Isolated To Component
+@rxFieldName-color: @gray-800; // Used In: rxForm, and rxCharacterCount
+@rxFieldName-symbol-color: @red-500; // Used In: rxForm
```
This might be good as `@app-required-symbol-color`

glynnis replied:
:+1:

CITguy commented on src/styles/vars.less:
```diff
@@ -320,8 +386,9 @@
 @progressbar-tertiary-background-color: @green-700;
 //** Font sizing for progress bar
-@progressbar-font-size-base: @appFontSize;
+@progressbar-font-size-base: @app-font-size;
```
This might be worth replacing with `@app-font-size` instead of reassigning to another variable.

glynnis replied:
:+1:

CITguy commented on src/styles/vars.less:
```diff
@@ -240,13 +300,15 @@
+ // Vars Are Not Isolated To Component
+@rxFieldName-color: @gray-800; // Used In: rxForm, and rxCharacterCount
+@rxFieldName-symbol-color: @red-500; // Used In: rxForm
+@rxFieldName-font-size: 12px; // Used In: rxForm
```
This might be good as `@app-label-font-size`

glynnis replied:
:+1:

CITguy commented on src/styles/vars.less:
```diff
@@ -209,17 +267,19 @@
 @rxCheckbox-background-color-selected: @rxForm-selected-accent-color;
 @rxCheckbox-border-color: @rxForm-input-border-color;
-@rxCheckbox-border-color-selected: @rxCheckbox-background-color-selected;
+@rxCheckbox-border-color-selected: @rxCheckbox-background-color-selected; // Not Used
```
ok to deprecate

CITguy commented on src/styles/vars.less:
```diff
@@ -257,7 +319,7 @@
 @rxSearchBox-background-disabled: @disabled-background-color;
 @rxSearchBox-border: 1px solid @rxForm-input-border-color;
-@rxSearchBox-border-radius: @rxForm-input-border-radius;
+@rxSearchBox-border-radius: @rxForm-input-border-radius; // Not Used
```
ok to deprecate

CITguy commented on src/styles/vars.less:
```diff
@@ -163,38 +219,40 @@
 @rxRadio-background-color: @rxForm-input-background-color;
 @rxRadio-color-selected: @rxForm-selected-accent-color;
-@rxRadio-color-disabled: @disabled-background-color;
+@rxRadio-color-disabled: @disabled-background-color; // Not Used
```
ok to deprecate

CITguy commented on src/styles/vars.less:
```diff
@@ -163,38 +219,40 @@
+ // TODO Vars Are Not Isolated To This Component's Files. Perhaps Make Them Global Vars with "app-" Prefix.
+@rxForm-font-size: 13px; // Used In: rxForm, and rxAccountInfo
+@rxForm-gutter: 16px; // best if divisible by 4 // Used In: rxForm
+@rxForm-section-min-width: 400px; // Not Used
```
ok to deprecate

CITguy commented on src/styles/vars.less:
```diff
@@ -163,38 +219,40 @@
+@rxForm-input-background-color: @white; // Used In: Vars, rxCheckbox, and rxMultiSelect
+@rxForm-input-min-height: 26px; // Used In: Vars, and rxForm
+@rxForm-input-width: 250px; // Used In: Vars
+@rxForm-input-number-width: 150px; // Not Used
```
ok to deprecate

ssalinas24 commented:
@glynnis and @CITguy  I have implemented the changes from your feedback, created the new global vars, implemented them where i didn't see big changes, and then created the following stories for the additional work to implement new global vars into the other components that may need them. Please proceed to review this PR.

https://jira.rax.io/browse/FRMW-593 
https://jira.rax.io/browse/FRMW-594
https://jira.rax.io/browse/FRMW-595
https://jira.rax.io/browse/FRMW-596
https://jira.rax.io/browse/FRMW-597
https://jira.rax.io/browse/FRMW-598
https://jira.rax.io/browse/FRMW-599
https://jira.rax.io/browse/FRMW-600
https://jira.rax.io/browse/FRMW-601
https://jira.rax.io/browse/FRMW-602
https://jira.rax.io/browse/FRMW-603
https://jira.rax.io/browse/FRMW-605
https://jira.rax.io/browse/FRMW-606
https://jira.rax.io/browse/FRMW-607
https://jira.rax.io/browse/FRMW-608
https://jira.rax.io/browse/FRMW-609

CITguy commented on src/styles/vars.less:
```diff
@@ -339,6 +377,10 @@
+/*
+ * rxPaginate
+ */
+@paginationColor: @gray-700;
```
Is this a new variable? If so, shouldn't it be kebab case?

ssalinas24 replied:
@CITguy  Darn, yep sure is. I meant to see if i could update. I will try to update ASAP.

CITguy commented:
LGTM

ssalinas24 commented:
@CITguy I updated that rxPaginate var that you pointed out, squashed, rebased, and submitted. Should be good now. Just kept it as a component var.

100stacks commented:
@ssalinas24, you might now hold the PR record for number of comments :wink: 

Needs rebase...LGTM :+1:
